### PR TITLE
feat: Add messaging for retention state exceeded

### LINKED ISSUE #24442: [IATR](M0) Failures page header component
Create a component to be the header of the new [failures page](https://github.com/cypress-io/cypress/issues/24441) that represents the chosen run and its metadata.

**NOTE:** M0 issues will be committed against a feature branch and not released 

### UI Design
* Will mimic the same header design as used on the Runs page in the app and a Run in the Cloud
* Element that will appear in header
  * Commit message
  * Run number
  * Run Status
  * Git branch name
  * Commit hash (shortened)
  * Committer username
  * Relative time to run
  * Run duration
  * "View in dashboard" link  that will link out to the run in the cloud via the CloudRun.url 

image

_Wireframe from Whimsical. Final design is in progress._

### Inputs
* CloudRun - Should be the same CloudRun type used on the Runs page
  * existing metadata that will be used in the header design
*  FailedSpecs[]: Spec

#### Resources
* [Clickup](https://app.clickup.com/t/18033298/PM-2546)
* [Whimsical](https://whimsical.com/q3-2022-solutions-CB1g7VnZeHsrVdz4gurF2o@AhRp651J2s3PQev2xg2uwbceF) - See V0 section

### LINKED ISSUE #25352: [IATR](M1.0) - Use new Cloud fields for isHidden determination
_Current behavior_

The App is determining when to show an empty state on the Debug page for when the results should be hidden by current looking at an `isHiddenByUsageLimits` field as well as for a run status of 'OVERLIMIT'.  This is happening in the `DebugPageDetails` component here:  https://github.com/cypress-io/cypress/blob/0a3ee16087283590f902b76e795bfea846658f03/packages/app/src/debug/DebugPageDetails.vue#L26-L30

There is also an additional state that is not currently being handled that is due to data retention limits on free plans.  There is currently no way in the App to determine this state.  Instead, the Cloud just does not return any items in the `specs` field.

_Desired Behavior_
The Cloud PR [[CYCLOUD-748] Updating CloudRun with isHidden and related fields](https://github.com/cypress-io/cypress-services/pull/5150/files) introduces additional fields on `CloudRun` to encapsulate the logic for determining if the run should be "hidden".
* `isHidden` is the one boolean to check instead of having to do conditional logic in the app
* `reasonsRunIsHidden` is a union type field that is either type `DataRetentionLimitExceeded` or `UsageLimitExceeded`.  These two types have additional information needed to construct the correct messaging shown in the App

~*Note* Until the above PR is merged, you can check out the branch `muaz/CYCLOUD-748-reasons-run-hidden` for the `cypress-services` project.~

*NOTE* The PR has been merged and these changes are now available on the main Cloud feature branch: `feature/CYCLOUD-665-IATR`

_Updates Needed_

Replace the conditional logic in `DebugPageDetails` to use the new `isHidden` field to show the `DebugOverLimit` .  Then use the data returned in the `reasonsRunIsHidden` to show the appropriate messaging for either the "usage limit" or the "data retention" scenarios.  This can either be done in the existing component or add an additional component for the "data retention" use case.

The `reasonRunIsHidden` types have fields for the potential variable portions of the messages:
* `dataRetentionDays` - https://github.com/cypress-io/cypress-services/pull/5150/files#diff-4d081635bf3fca6e0a14c54b8943221997c052f52ec542165e41b7521596524cR1094-R1099
* `monthlyTests` - https://github.com/cypress-io/cypress-services/pull/5150/files#diff-4d081635bf3fca6e0a14c54b8943221997c052f52ec542165e41b7521596524cR1299-R1304

Figma designs:
* Over limit
  * [Is admin](https://www.figma.com/file/PAOozuCxtfetqP8fUVvg0y/branch/QkCp2F69X0qyLo7qnYQ67e/Cypress-App?node-id=21687%3A213116&t=jaAfdAZZXaA9I3kf-1)
  * [Not admin](https://www.figma.com/file/PAOozuCxtfetqP8fUVvg0y/branch/QkCp2F69X0qyLo7qnYQ67e/Cypress-App?node-id=21687%3A213116&t=jaAfdAZZXaA9I3kf-1)
* Data retention
  * [Is admin](https://www.figma.com/file/PAOozuCxtfetqP8fUVvg0y/branch/QkCp2F69X0qyLo7qnYQ67e/Cypress-App?node-id=23388%3A222416&t=jaAfdAZZXaA9I3kf-1)
  * [Not admin](https://www.figma.com/file/PAOozuCxtfetqP8fUVvg0y/branch/QkCp2F69X0qyLo7qnYQ67e/Cypress-App?node-id=23388%3A222416&t=jaAfdAZZXaA9I3kf-1)

[CYCLOUD-748]: https://cypress-io.atlassian.net/browse/CYCLOUD-748?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CYCLOUD-748]: https://cypress-io.atlassian.net/browse/CYCLOUD-748?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### PR DESCRIPTION
**Pull request by mike-plummer**
- Closes #25352  

### User facing changelog
No changelog entry, merging to feature branch

### Additional details

Adding handling for new "Data Retention" messaging state when an OVERLIMIT is reported. Extending messaging for "usage" overlimits by adding test recording limit value to message.

See #25352 for details

### Steps to test
See `DebugRunStates.cy.tsx` for new states, Percy snapshots generated for all four permutations (Retention User, Retention Admin, Usage User, Usage Admin)

### How has the user experience changed?

| State | Screenshot |
|-----|-----|
| Usage - User | Screen Shot 2023-01-04 at 1 35 54 PM |
| Usage - Admin | Screen Shot 2023-01-04 at 1 36 01 PM |
| Retention - User | Screen Shot 2023-01-04 at 1 36 09 PM |
| Retention - Admin | Screen Shot 2023-01-04 at 1 36 17 PM |

### PR Tasks
- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/app/src/debug/DebugPageDetails.vue [hunk 1]
```diff
@@ -24,9 +24,11 @@
       :ci="ci"
     />
     <DebugOverLimit
-      v-else-if="status === 'OVERLIMIT' || isHiddenByUsageLimits"
+      v-else-if="status === 'OVERLIMIT' || isHidden"
```
Comment (warrensplayer) [inline packages/app/src/debug/DebugPageDetails.vue]: The `isHidden` flag will be true when the status is `OVERLIMIT`, so no need to include that check here.  Just check the `isHidden` flag

File: packages/app/src/debug/DebugPageDetails.vue [hunk 2]
```diff
@@ -24,9 +24,11 @@
...
+      :over-limit-reason="reasonsRunIsHidden[0] || null"
```
Comment (warrensplayer) [inline packages/app/src/debug/DebugPageDetails.vue]: I did not realize that this was an array, but it makes sense that it could be.  I think the decision to pull the first reason to show makes sense.  However, I think it would be good to leave a comment explaining why this is happening.  Also, I think it would make sense to push this decision into the `DebugOverLimit` component.

TLDR:  Pass the entire array into the component and then put a comment where it is selecting the first value in the array to show that says something like:  "Choosing to only display the first over limit reason."

**Review by warrensplayer — CHANGES_REQUESTED** [review]
Looks great. Only two minor updates.

File: packages/app/src/debug/DebugPageDetails.vue (hunk 2 above)
Comment (mike-plummer) [reply packages/app/src/debug/DebugPageDetails.vue]: Shoot, meant to refactor this and lost track of it. Will look at pushing this down into the lower component.

File: packages/app/src/debug/DebugOverLimit.vue [hunk 3]
```diff
@@ -27,15 +28,56 @@ import { useI18n } from '@cy/i18n'
...
+
+const iconClasses = computed(() => {
+  return [
+    'icon-dark-gray-500',
+    {
+      'icon-dark-secondary-jade-400': !isRetentionLimit(props.overLimitReason),
+      'icon-light-secondary-jade-200': !isRetentionLimit(props.overLimitReason),
+      'icon-dark-secondary-orange-400': isRetentionLimit(props.overLimitReason),
+      'icon-light-secondary-orange-200': isRetentionLimit(props.overLimitReason),
+    },
```
Comment (ZachJW34) [inline packages/app/src/debug/DebugOverLimit.vue]:
```suggestion
   isRetentionLimit(props.overLimitReason)
      ? 'icon-dark-secondary-orange-400 icon-light-secondary-orange-200'
      : 'icon-dark-secondary-jade-400 icon-light-secondary-jade-200',
```

Nit: it's nice to have the classes grouped together

File: packages/app/src/debug/DebugOverLimit.vue [hunk 4]
```diff
@@ -27,15 +28,65 @@ import { useI18n } from '@cy/i18n'
...
+    return {
+      title: t('debugPage.overLimit.retentionExceededTitle'),
+      message: t('debugPage.overLimit.retentionExceededMessage', { limitDays: overLimitReason.value.dataRetentionDays, runAgeDays: props.runAgeDays }),
+      actionLabel: isPlanAdmin.value ? t('debugPage.overLimit.upgradePlan') : t('debugPage.overLimit.contactAdmin'),
+    }
+  }
+
+  return {
+    title: t('debugPage.overLimit.usageExceededTitle'),
+    message: isPlanAdmin.value ? t('debugPage.overLimit.usageExceededAdminMessage', { numberTests: overLimitReason.value?.monthlyTests || 0 }) : t('debugPage.overLimit.usageExceededUserMessage', { numberTests: overLimitReason.value?.monthlyTests || 0 }),
```
Comment (ZachJW34) [inline packages/app/src/debug/DebugOverLimit.vue]: Nit: Creating a variable for `numberTests` and using object shorthand here makes this easier to read.

File: packages/app/src/debug/DebugPageDetails.vue (hunk 1 above)
Comment (ZachJW34) [reply packages/app/src/debug/DebugPageDetails.vue]: Bumping in case this was overlooked.

File: packages/app/src/debug/DebugOverLimit.vue [hunk 5]
```diff
@@ -27,15 +28,65 @@ import { useI18n } from '@cy/i18n'
 import { getUrlWithParams } from '@packages/frontend-shared/src/utils/getUrlWithParams'
 import { computed } from '@vue/reactivity'
 
+export type DataRetentionLimitExceeded = { __typename: 'DataRetentionLimitExceeded', dataRetentionDays: number | null }
+
+export type UsageLimitExceeded = | { __typename: 'UsageLimitExceeded', monthlyTests: number | null }
+
+export type CloudRunHidingReason = DataRetentionLimitExceeded | UsageLimitExceeded
```
Comment (ZachJW34) [inline packages/app/src/debug/DebugOverLimit.vue]: Can we create a fragment out of this and derive the types from it?

```
gql`
fragment DebugReasonsRunIsHidden on CloudRun {
  id
  reasonsRunIsHidden {
    __typename
    ... on DataRetentionLimitExceeded {
      dataRetentionDays
    }
    ... on UsageLimitExceeded {
      monthlyTests
    }
  }
}
`

type CLoudRunHidingReason = DebugReasonsRunIsHidden['reasonsRunIsHidden']
```

That way we don't have to duplicate types.

**Review by ZachJW34 — APPROVED** [review]
Mostly nits, going to approve assuming that @warrensplayer comment on the `isHidden` will be addressed.

**Review by warrensplayer — APPROVED** [review]
Looks good!
